router/router_test.go

```diff
@@ -108,3 +104,129 @@ func TestRingpopRouterMakeRemoteClientCached(t *testing.T) {
 	router.GetClient("hello")
 	cf.AssertNumberOfCalls(t, "MakeRemoteClient", 1)
 }
+
+func TestRingpopRouterRemoveExistingClient(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.removeClient(hostport)
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveNonExistingClient(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	assert.Equal(t, nil, router.clientCache[hostport])
+	router.removeClient(hostport)
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveClientOnSwimFaultyEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.HandleEvent(swim.MemberlistChangesReceivedEvent{
+		Changes: []swim.Change{
+			swim.Change{
+				Address: hostport,
+				Status:  swim.Faulty,
+			},
+		},
+	})
+
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveClientOnSwimLeaveEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.HandleEvent(swim.MemberlistChangesReceivedEvent{
+		Changes: []swim.Change{
+			swim.Change{
+				Address: hostport,
+				Status:  swim.Leave,
+			},
+		},
+	})
+
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterNotRemoveClientOnSwimSuspectEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
```

Comment:
This approach of directly manipulating the map in the tests gives me a bit of pause.  I'm not saying that we *definitely* shouldn't do this (sometimes it's a favorable tradeoff), but since the tests are in the same package, I wonder whether a method like getClientNoCreate or clientExists or something might not give us a little encapsulation that would make the tests more resilient to refactoring.  What do you think?